### PR TITLE
APM metrics filters

### PR DESCRIPTION
**Pull request by ekremsekerci**
Add filtering capability for metrics ingestion to third-party APMs

APM_METRICS_FILTER_ALLOW and APM_METRICS_FILTER_DENY environment variables are introduced.

Also in telegraf configuration round_interval set to true for internal metrics

### CODE REVIEW

File: buildpack/telemetry/metrics.py [hunk 1]
```diff
@@ -55,9 +55,37 @@
...
+    it accepts/denies every metric since we us type as `nameStartsWith`.
+    To prevent breaking the functionality because of this, we need to make sure
+    that we pass empty string to the registry filters only if it's intentional.
+    So, we strip the leading and trailing commas. Additionally we remove all
+     the white spaces to prevent any unintentional mistakes.
+    """
+    env_var_value = os.getenv(env_var_name)
+    if env_var_value is None:
+        return []
+    return env_var_value.replace(" ", "").strip(",").split(",")
```
Comment (pappachino) [inline buildpack/telemetry/metrics.py:71]: Not sure why we need the additional `strip` 🤔 

This should work right ?
```suggestion
    return env_var_value.replace(" ", "").split(",")
```

**Review by pappachino — CHANGES_REQUESTED** [review]
Can we add some unit tests similar to what we have [here](https://github.com/mendix/cf-mendix-buildpack/blob/master/tests/unit/test_micrometer_metrics.py#L65)

File: buildpack/telemetry/metrics.py (hunk 1 above)
Comment (ekremsekerci) [reply buildpack/telemetry/metrics.py:71]: If the there is a leading/trailing comma then split creates empty string in the resulting list which breaks our logic since the empty string means all metrics in this case. 
image

File: buildpack/telemetry/metrics.py [hunk 2]
```diff
@@ -55,9 +55,37 @@
     ],
 }
 
+
+def get_apm_metrics_filter(env_var_name):
+    """
+    If we use empty string ("") in the filters that we use for statsd registry,
+    it accepts/denies every metric since we us type as `nameStartsWith`.
```
[the remark this answers is not in the document]
Comment (ekremsekerci) [reply buildpack/telemetry/metrics.py]: Either I'll add a similar explanation to the docs or refer to this.

Fixed the typo, thanks

File: buildpack/telemetry/metrics.py [hunk 3]
```diff
@@ -55,9 +55,37 @@
...
+            "type": "nameStartsWith",
+            "result": "accept",
+            "values": get_apm_metrics_filter("APM_METRICS_FILTER_ALLOW"),
+        },
+        {
+            "type": "nameStartsWith",
+            "result": "deny",
+            "values": get_apm_metrics_filter("APM_METRICS_FILTER_DENY"),
+        },
+    ],
```
[the remark this answers is not in the document]
Comment (ekremsekerci) [reply buildpack/telemetry/metrics.py:88]: Yeah, then it's not effective. Will add this case to the peer testing steps, thanks 👍

**Comment by ekremsekerci** [comment]
> Can we add some unit tests similar to what we have [here](https://github.com/mendix/cf-mendix-buildpack/blob/master/tests/unit/test_micrometer_metrics.py#L65)

The unit tests we have there are to determine which registries will be used under which conditions. With my changes, I don't create any other possibility. I'm just playing with the existing statsd registry.

Some unit tests can be added to test the method `get_apm_metrics_filter`. However, those tests mostly will be like testing `replace`, `strip`, and `split` functionalities so I was not so sure if adding those tests add much value. If you think it adds value I can add?

File: buildpack/telemetry/metrics.py (hunk 1 above)
Comment (pappachino) [reply buildpack/telemetry/metrics.py:71]: Ah okay, got it. It's a bit of over-optimization though as to why we only care about the comma 😄

**Comment by pappachino** [comment]
> > Can we add some unit tests similar to what we have [here](https://github.com/mendix/cf-mendix-buildpack/blob/master/tests/unit/test_micrometer_metrics.py#L65)
> 
> The unit tests we have there are to determine which registries will be used under which conditions. With my changes, I don't create any other possibility. I'm just playing with the existing statsd registry.
> 
> Some unit tests can be added to test the method `get_apm_metrics_filter`. However, those tests mostly will be like testing `replace`, `strip`, and `split` functionalities so I was not so sure if adding those tests add much value. If you think it adds value I can add?

Yeah, mainly to ensure we are applying the expected filters when we set the ALLOW and DENY env vars. Better to add some unit-tests to avoid any future regressions.

File: buildpack/telemetry/metrics.py [hunk 4]
```diff
@@ -55,9 +55,37 @@
...
+    env_var_value = os.getenv(env_var_name)
+    if env_var_value is None:
+        return []
+    return env_var_value.replace(" ", "").strip(",").split(",")
+
+
 STATSD_REGISTRY = {
     "type": "statsd",
     "settings": {"port": datadog.get_statsd_port()},
+    "filters": [
```
Comment (pappachino) [inline buildpack/telemetry/metrics.py:77]: Can we log the final filter values only if the user has provided any of the envvars? Would help us debug if the user has entered unexpected patterns.

**Review by pappachino — CHANGES_REQUESTED** [review]

File: buildpack/telemetry/metrics.py (hunk 1 above)
Comment (ekremsekerci) [reply buildpack/telemetry/metrics.py:71]: We only care about comma because we split by comma :) We'll be waiting for comma separated values as input and the logic breaks if there is a leading/trailing comma in the input. So basically, I'm trying to prevent unintentional errors

File: buildpack/telemetry/metrics.py (hunk 4 above)
Comment (pappachino) [reply buildpack/telemetry/metrics.py:79]: I think `INFO` is fine too, as it should show up only once during startup right? Else we need to ask them to enable `DEBUG` to figure out any issue.

File: etc/telegraf/telegraf.toml.j2 [hunk 5]
```diff
@@ -29,6 +29,7 @@
     commands = ["echo {{ is_enabled }}"]
     name_override = "integration_usage"
     interval = "6h"
+    round_interval = false
```
Comment (pappachino) [inline etc/telegraf/telegraf.toml.j2:32]: A note on why we overriding this would be helpful

File: README.md [hunk 6]
```diff
@@ -820,6 +820,30 @@ Example (1000 loglines/second):
 cf set-env <YOUR_APP> LOG_RATELIMIT '1000'
 ```
 
+### Custom Runtime Metrics filtering
+
+For the third-party integrations explained above, in addition to the metrics collected by the agents, custom runtime metrics are provided via telegraf.
+This configuration also has a filtering mechanism that allows users to specify metrics they allow or deny for the vendor they are using.
+To be able to filter the emission of custom runtime metrics, users should provide list of prefixes of metrics they need to allow/deny using the environment variables listed below.
```
Comment (msvolenski) [inline README.md:827]: "To filter the emission of custom runtime metrics, users should provide a list of prefixes of the metrics they want to allow/deny using the environment variables listed below."

Just a minor rewriting.

File: README.md [hunk 7]
```diff
@@ -820,6 +820,30 @@ Example (1000 loglines/second):
...
+
+Note: Custom database metrics cannot be filtered by name, to turn them off `APPMETRICS_INCLUDE_DB` environment variable should be set to false.
```
Comment (msvolenski) [inline README.md:829]: "Note: Custom database metrics cannot be filtered by name, to turn them off, the `APPMETRICS_INCLUDE_DB` environment variable should be set to false."

Just a minor rewriting.

File: README.md [hunk 8]
```diff
@@ -820,6 +820,30 @@ Example (1000 loglines/second):
...
+This configuration also has a filtering mechanism that allows users to specify metrics they allow or deny for the vendor they are using.
+To be able to filter the emission of custom runtime metrics, users should provide list of prefixes of metrics they need to allow/deny using the environment variables listed below.
+
+Note: Custom database metrics cannot be filtered by name, to turn them off `APPMETRICS_INCLUDE_DB` environment variable should be set to false. 
+
+#### APM_METRICS_FILTER_DENY
+
+List of prefixes for the metrics to be denied. 
+
+For example to deny all metrics starting with **jetty** or **mx.runtime** the env var should be set to `jetty,mx.runtime`.
```
Comment (msvolenski) [inline README.md:835]: "For example, to deny all metrics starting with **jetty** or **mx.runtime**, the environment variable should be set to `jetty,mx.runtime`."

Just a minor rewriting.

File: README.md [hunk 9]
```diff
@@ -820,6 +820,30 @@ Example (1000 loglines/second):
...
+
+To deny all the custom metrics, the environment variable should be set without any value.
+
+
+
+#### APM_METRICS_FILTER_ALLOW
+
+List of prefixes for the metrics to be allowed. By default, all metrics are allowed even if they are not specified via this env var, so it makes sense to use this environment variable in combination with the `APM_METRICS_FILTER_DENY`.
```
Comment (msvolenski) [inline README.md:843]: "List of prefixes for the metrics to be allowed. By default, all metrics are allowed, even if they are not specified via this env var, so it makes sense to use this environment variable in combination with the `APM_METRICS_FILTER_DENY`."

Just a minor rewriting.

File: README.md [hunk 10]
```diff
@@ -820,6 +820,30 @@ Example (1000 loglines/second):
...
+
+For example to deny all metrics except the session metrics, `APM_METRICS_FILTER_DENY` should be used without any value as explained before and `APM_METRICS_FILTER_ALLOW` should be set to `mx.runtime.stats.sessions`.
```
Comment (msvolenski) [inline README.md:845]: "For example, to deny all metrics except the session metrics, `APM_METRICS_FILTER_DENY` should be used without any value as explained before, and `APM_METRICS_FILTER_ALLOW` should be set to `mx.runtime.stats.sessions`."

Just a minor rewriting.

File: buildpack/telemetry/metrics.py [hunk 11]
```diff
@@ -186,21 +123,119 @@ def configure_metrics_registry(m2ee):
...
+    if env_var_value is None:
+        return []
+    apm_filter = env_var_value.replace(" ", "").strip(",").split(",")
+    logging.info("APM metrics filter for %s is: %s", env_var_value, apm_filter)
+    return apm_filter
+
+
+def get_freeapps_registry():
+    # For freeapps we push only the session & login metrics
+    return [
```
Comment (msvolenski) [inline buildpack/telemetry/metrics.py:216]: I know you didn't change that, but I would like to suggest that we return a dictionary here and add the list in the method that calls it, as it's done for the others.

File: tests/unit/test_micrometer_metrics.py [hunk 12]
```diff
@@ -97,12 +97,43 @@ def test_paidapps_metrics_registry_statsd(self, is_enabled):
...
+            {
+                "PROFILE": "some-random-mx-profile",
+                "APM_METRICS_FILTER_ALLOW": "allowed_metric",
+                "APM_METRICS_FILTER_DENY": "denied_metric",
+            },
+        ):
+            result = metrics.configure_metrics_registry(Mock())
+            metrics_registries = sorted(result, key=itemgetter("type"))
+
+            # influx registry shouldn't the filters
```
Comment (msvolenski) [inline tests/unit/test_micrometer_metrics.py:113]: "[...] shouldn't have [...]"

File: etc/telegraf/telegraf.toml.j2 (hunk 5 above)
Comment (ekremsekerci) [reply etc/telegraf/telegraf.toml.j2]: removed this already

File: README.md [hunk 13]
```diff
@@ -820,6 +820,30 @@ Example (1000 loglines/second):
...
+
+Note: Custom database metrics cannot be filtered by name, to turn them off `APPMETRICS_INCLUDE_DB` environment variable should be set to false. 
+
+#### APM_METRICS_FILTER_DENY
+
+List of prefixes for the metrics to be denied. 
+
+For example to deny all metrics starting with **jetty** or **mx.runtime** the env var should be set to `jetty,mx.runtime`.
+
+To deny all the custom metrics, the environment variable should be set without any value.
```
[the remark this answers is not in the document]
Comment (ekremsekerci) [reply README.md]: added

File: README.md [hunk 14]
```diff
@@ -820,6 +820,30 @@ Example (1000 loglines/second):
...
+
+
+
+#### APM_METRICS_FILTER_ALLOW
```
[the remark this answers is not in the document]
Comment (ekremsekerci) [reply README.md:847]: added

File: README.md (hunk 6 above)
Comment (ekremsekerci) [reply README.md]: updated

File: README.md (hunk 7 above)
Comment (ekremsekerci) [reply README.md]: updated

File: README.md (hunk 8 above)
Comment (ekremsekerci) [reply README.md]: updated

File: README.md (hunk 9 above)
Comment (ekremsekerci) [reply README.md]: updated

File: README.md (hunk 10 above)
Comment (ekremsekerci) [reply README.md]: updated

File: buildpack/telemetry/metrics.py (hunk 11 above)
Comment (ekremsekerci) [reply buildpack/telemetry/metrics.py]: updated, thanks for the suggestion 👍

File: tests/unit/test_micrometer_metrics.py (hunk 12 above)
Comment (ekremsekerci) [reply tests/unit/test_micrometer_metrics.py]: updated

File: README.md [hunk 15]
```diff
@@ -824,36 +824,31 @@ cf set-env <YOUR_APP> LOG_RATELIMIT '1000'
 
 For the third-party integrations explained above, in addition to the metrics collected by the agents, custom runtime metrics are provided via telegraf.
 This configuration also has a filtering mechanism that allows users to specify metrics they allow or deny for the vendor they are using.
-To filter the emission of custom runtime metrics, users should provide a list of prefixes of the metrics they want to allow/deny using the environment variables listed below.
+To filter the ingestion of custom runtime metrics to third party APMs, users should provide a list of prefixes of the metrics they want to allow/deny using the environment variables listed below.
```
Comment (pappachino) [inline README.md:827]: 💯 Yeah emission for some reason didn't sound right. This is good.

File: README.md [hunk 16]
```diff
@@ -824,36 +824,31 @@ cf set-env <YOUR_APP> LOG_RATELIMIT '1000'
...
 Note: Custom database metrics cannot be filtered by name, to turn them off, the `APPMETRICS_INCLUDE_DB` environment variable should be set to false. 
 
-#### APM_METRICS_FILTER_DENY
+#### APM_METRICS_FILTER_ALLOW
 
-List of prefixes for the metrics to be denied. 
+List of prefixes for the metrics to be allowed. By default, all metrics are allowed, even if they are not specified via this env var.
 
-For example, to deny all metrics starting with jetty or mx.runtime, the environment variable should be set to `jetty,mx.runtime`:
+For example, to deny all metrics except the session metrics, `APM_METRICS_FILTER_ALLOW` should be set to `mx.runtime.stats.sessions`:
```
Comment (pappachino) [inline README.md:835]: As we're in the `ALLOW` section, let's rephrase it to
```suggestion
For example, to allow only the session metrics, `APM_METRICS_FILTER_ALLOW` should be set to `mx.runtime.stats.sessions`:
```

File: README.md [hunk 17]
```diff
@@ -824,36 +824,31 @@ cf set-env <YOUR_APP> LOG_RATELIMIT '1000'
...
-To filter the emission of custom runtime metrics, users should provide a list of prefixes of the metrics they want to allow/deny using the environment variables listed below.
+To filter the ingestion of custom runtime metrics to third party APMs, users should provide a list of prefixes of the metrics they want to allow/deny using the environment variables listed below.
 
 Note: Custom database metrics cannot be filtered by name, to turn them off, the `APPMETRICS_INCLUDE_DB` environment variable should be set to false. 
 
-#### APM_METRICS_FILTER_DENY
+#### APM_METRICS_FILTER_ALLOW
 
-List of prefixes for the metrics to be denied. 
+List of prefixes for the metrics to be allowed. By default, all metrics are allowed, even if they are not specified via this env var.
```
Comment (pappachino) [inline README.md:833]: Let's rephrase it to
```suggestion
Comma-separated list of prefixes for the metrics to be allowed. By default, all metrics are allowed, even if they are not specified via this env var.
```

File: README.md [hunk 18]
```diff
@@ -824,36 +824,31 @@ cf set-env <YOUR_APP> LOG_RATELIMIT '1000'
...
-cf set-env <YOUR_APP> APM_METRICS_FILTER_DENY ''
+cf set-env <YOUR_APP> APM_METRICS_FILTER_ALLOW 'mx.runtime.stats.sessions'
 ```
 
+#### APM_METRICS_FILTER_DENY
 
-#### APM_METRICS_FILTER_ALLOW
-
-List of prefixes for the metrics to be allowed. By default, all metrics are allowed, even if they are not specified via this env var, so it makes sense to use this environment variable in combination with the `APM_METRICS_FILTER_DENY`.
+List of prefixes for the metrics to be denied.
```
Comment (pappachino) [inline README.md:843]:
```suggestion
Comma-separated list of prefixes for the metrics to be denied. 
```

File: buildpack/telemetry/metrics.py [hunk 19]
```diff
@@ -176,25 +176,41 @@ def get_influx_registry():
 
 
 def get_statsd_registry():
+    allowed_metrics = os.getenv("APM_METRICS_FILTER_ALLOW")
+    denied_metrics = os.getenv("APM_METRICS_FILTER_DENY")
+
+    if allowed_metrics and not denied_metrics:
```
Comment (pappachino) [inline buildpack/telemetry/metrics.py]:
```suggestion
    if allowed_metrics and denied_metrics is not None:
```
More of a nitpick. Do we need to support falsey prefixes like `0` or `false`

File: buildpack/telemetry/metrics.py [hunk 20]
```diff
@@ -176,25 +176,41 @@ def get_influx_registry():
 
 
 def get_statsd_registry():
+    allowed_metrics = os.getenv("APM_METRICS_FILTER_ALLOW")
```
Comment (pappachino) [inline buildpack/telemetry/metrics.py]: Let's keep this allow and deny logic outside of `get_statsd_registry` method, as it is not tied to the registry functionality and can be re-used if needed.

File: buildpack/telemetry/metrics.py (hunk 19 above)
Comment (ekremsekerci) [reply buildpack/telemetry/metrics.py]: Here we're looking for (denied_metrics is None), but yeah for being more explicit, I'm refactoring my version to this. 

I think we don't need to validate the content here because it can easily be understood by users and fixed even if they do some mistakes on the first attempts

File: README.md (hunk 16 above)
Comment (ekremsekerci) [reply README.md]: applied

File: README.md (hunk 18 above)
Comment (ekremsekerci) [reply README.md]: applied

File: buildpack/telemetry/metrics.py (hunk 20 above)
Comment (ekremsekerci) [reply buildpack/telemetry/metrics.py]: isolated the logic in a separate method now. get_statsd_registry can be called with empty lists if we need to use it anywhere else without filters

File: README.md [hunk 21]
```diff
@@ -848,6 +848,16 @@ For example, to deny all metrics starting with jetty or mx.runtime, the environm
...
 ```
 
+#### APM_METRICS_FILTER_DENY_ALL
+
+If this environment variable is set to `true`, all metrics will be denied regardless of values of `APM_METRICS_FILTER_ALLOW` and `APM_METRICS_FILTER_DENY`. 
+
+Be aware that `APPMETRICS_INCLUDE_DB` still needs to be set to false to turn off the database metrics.
+
+```shell
+cf set-env <YOUR_APP> APM_METRICS_FILTER_DENY 'jetty,mx.runtime'
```
[the remark this answers is not in the document]
Comment (ekremsekerci) [reply README.md]: updated

File: buildpack/telemetry/telegraf.py [hunk 22]
```diff
@@ -75,6 +75,8 @@ def _get_config_file_path(version):
 
 
 def include_db_metrics():
+    if metrics.deny_all_apm_metrics():
```
Comment (msvolenski) [inline buildpack/telemetry/telegraf.py:78]: This could be put in the same condition below, but I understand keeping the free apps in a separate logic.
So I'm fine with both.

Comment (ekremsekerci) [reply buildpack/telemetry/telegraf.py:78]: yeah, I just wanted to make it very explicit and separate in the beginning since this env var precedes all the other logic

**Review by pappachino — APPROVED** [review]
